core/src/main/java/com/linecorp/armeria/common/stream/StreamMessage.java

```diff
@@ -752,6 +754,11 @@ default CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionO
         requireNonNull(executor, "executor");
         requireNonNull(options, "options");
         final StreamMessageCollector<T> collector = new StreamMessageCollector<>(options);
+        if (!containsNotifyCancellation(options)) {
```

Comment:
Before setting this PR as ready for review, it would be easier to review if you explain why this change is needed in the PR description

Reply:
`StreamMessageCollector` collects the published elements until `onComplete()` or `onError()` is called. It generally worked but I found a corner case where `Subscription.cancel()` made `collector.collect()` incomplete forever. 

You can reproduce it by disabling this block and then running it below.
https://github.com/line/armeria/blob/a870edf92a5840b4e0e4cbbfeed562e7c8174b4f/core/src/test/java/com/linecorp/armeria/common/stream/StreamMessageCollectingTest.java#L168

I believe it would make more sense to specify `NOTIFY_CANCELLATION` for `collect()` method because it is used to fully consume the upstream data instead of partial ones.
`CancelledSubscriptionException` will make the collecting future completed exceptionally.

Reply:
For context, previously, the custom `collect()` implementation in `FilteredStreamMessage` hooked a `Subscription.cancel()` event and completed the collecting future with the accumulated items so far.

Reply:
I see, I understood the issue now. Agree with the approach 👍